Begin implementation of `ot_keymgr` for Earlgrey

AlexJones0 opened the pull request:
This PR is the first of a series of 4 PRs to implement OpenTitan's `keymgr` IP, as used in Earlgrey, split to ease review. You can see all the relevant `keymgr` commits (including this PR) [here](https://github.com/lowRISC/qemu/compare/ot-earlgrey-9.2.0...AlexJones0:qemu:eg-keymgr), which is functionally complete, though note that small details in future commits may still be WIP.

The `keymgr` is quite similar to the existing `ot_keymgr_dpe`, meaning that a lot of the logic is similar and the design of the block follows a similar structure. Despite this, their design is different - `keymgr` has unique states for each key stage and multiple CDIs, whereas `keymgr_dpe` has no CDIs but many slots with different boot stages, and only 1 generic FSM key state. There has also been a fair amount of RTL divergence between the two IPs, meaning that many specifics of the implementation are unique to Earlgrey. Thus, these PRs create the `keymgr` from scratch, attempting to note what is the same as / changed from `keymgr_dpe` in the commit messages where relevant for review.

One note: for now I have opted to use separate `ot_keymgr` traces from the existing `ot_keymgr_dpe` traces, because in future PRs the blocks diverge a bit further and use different traces for different things (and `keymgr_dpe` could perhaps diverge further in the future). However, I think both blocks could probably just use generic `ot_keymgr` traces with redundant traces left unused for each block. 

This PR adds a lot of the initial description for the `keymgr`, and sets up initial functionality like register reads/writes, IRQs, alerts. This stubbed keymgr is then connected to Earlgrey. Future PRs will implement the main FSM and `keymgr` operation, but they are separated from this PR to ease review.

See the commit messages for more details about each change.

AlexJones0 committed:
[ot] hw/opentitan: ot_keymgr: Add initial stubbed keymgr implementation

This adds a very basic stubbed keymgr implementation with all of the
different registers and fields (and some constant pkg values) defined.
This is entirely stubbed at the moment - all writes and reads to
registers are unimplemented, as are register reset values and any of
the internal logic. The intention is to build upon the keymgr
implementation in further commits.

Signed-off-by: Alex Jones <alex.jones@lowrisc.org>

AlexJones0 committed:
[ot] hw/riscv: ot_earlgrey: Connect stubbed keymgr

Replace the default unimplemented keymgr device in the Earlgrey build
with the stubbed keymgr device, which will be iteratively developed.
Now that the stubbed keymgr has its seed properties and IRQ & alert
updates it should be at least as useful as the unimplemented device.

Signed-off-by: Alex Jones <alex.jones@lowrisc.org>

AlexJones0 committed:
[ot] hw/opentitan: ot_keymgr: Add register reset values

Signed-off-by: Alex Jones <alex.jones@lowrisc.org>

AlexJones0 committed:
[ot] hw/opentitan: ot_keymgr: Add IRQ and alert updates

In the keymgr, alerts from the ALERT_TEST register and recoverable
operation error alerts are transient (they signal once and then
de-assert), whilst fatal fault alerts are latched (they stay
signaled). This logic is implemented in the `update_alerts` function.

Signed-off-by: Alex Jones <alex.jones@lowrisc.org>

AlexJones0 committed:
[ot] hw/opentitan: ot_keymgr: Implement basic reg read/writes

Implements the logic for keymgr register reads and writes, mostly
limited to write masking, RW0C/RW1C/RO/WO behaviour, REGWENs and
shadowed registers. A small bit of additional logic is added to hook
up the interrupt/alert state/test registers to propagate IRQs and
alert signals.

Some logic is also implemented to load the salt and sealing keys into
separate byte arrays, since these will be needed for later parts of the
implementation. The RC (read-clear) software share outputs are left
completely unimplemented for now as these will be added later along with
their implementation.

Signed-off-by: Alex Jones <alex.jones@lowrisc.org>

AlexJones0 committed:
[ot] scripts/opentitan: cfggen.py: Extract RndCnstCdi from Keymgr HJSON

The OT keymgr HJSON features a parameter "RndCnstCdi" containing the
compile-time random bits to use for the generation seed when no CDI is
selected. This is like "RndCnstNoneSeed", but used when there is no CDI
selected instead of no destination selected.

Unlike the other compile-time random seeds, this is not suffixed with
"Seed" and thus is not extracted by the existing regex.

This commit adds a special case for this in `cfggen.py` to extract this
attribute, and appropriately rename it to a `cdi_seed` property instead
of `cdi` so that it converges with the other seed config parameters.

Signed-off-by: Alex Jones <alex.jones@lowrisc.org>

AlexJones0 committed:
[ot] hw/opentitan: ot_keymgr: Add seed constant properties

Add properties for loading compile-time generation seeds for the keymgr
from the OpenTitan configuration file. This includes defining the
different seeds that are present in the keymgr, which diverge from those
present in the keymgr_dpe due to the differences in the design.

Signed-off-by: Alex Jones <alex.jones@lowrisc.org>

rivos-eblot commented on scripts/opentitan/cfggen.py:
```diff
@@ -135,6 +135,11 @@ def load_top_config(self, toppath: str) -> None:
                 self._load_top_values(module, self._keymgr, False,
                                       r'RndCnst((?:.*)Seed)')
+                # The CDI keymgr seed does not match `RndCnst.*Seed`
+                self._load_top_values(module, self._keymgr, False,
```
I implemented it but never tested it: did you try:

`self._load_top_values(module, self._keymgr, False,  r'RndCnst((?:.*)Seed)',  r'RndCnst(Cdi)')` or even a combined regex with `|`?

AlexJones0 replied:
I did and it worked, but I separated it so I could attach a separate comment to make the code easier to follow (to understand why we special case this). I'm happy to change it though.

AlexJones0 replied:
I've changed it now (but kept the comment). The formatting becomes a bit verbose but otherwise from testing it seems to work just as well.

rivos-eblot replied:
Ok thanks. It seems fine with the split-line syntax.

rivos-eblot commented on hw/opentitan/ot_keymgr.c:
```diff
@@ -0,0 +1,1015 @@
+ */
+
+#include "qemu/osdep.h"
+#include <assert.h>
```
It shouldn't be required as `g_assert` should be used

AlexJones0 replied:
Indeed, this is also present in `keymgr_dpe` and got pasted over - I'll add commits dropping those from `keymgr_dpe` to this PR.

rivos-eblot replied:
Oops, sorry. Maybe this comes from my stupid "smart" editor than can't help but adding useless headers...

rivos-eblot commented on hw/opentitan/ot_keymgr.c:
```diff
@@ -0,0 +1,1015 @@
+
+#include "qemu/osdep.h"
+#include <assert.h>
+#include <stdint.h>
```
It should not be required, as "osdep.h" takes care of all system-wide header files.

AlexJones0 replied:
Same as above, thanks for the info.

rivos-eblot reviewed (dismissed):
I do not know the keymgr, maybe @loiclefort has better knowledge on this EarlGrey version.

AlexJones0 committed:
[ot] hw/opentitan: ot_keymgr_dpe: Drop unused imports

"assert.h" is not used as we use QEMU's `g_assert` instead.
"stdint.h" should not be included as QEMU's "osdep.h" takes care of all
system-wide header files for us.

Signed-off-by: Alex Jones <alex.jones@lowrisc.org>

rivos-eblot reviewed (commented):
LGTM (but I do not know the keymgr, so leaving knowledgeable people to approve the PR)

jwnrt commented on hw/opentitan/ot_keymgr.c:
```diff
@@ -564,8 +583,58 @@ static void ot_keymgr_write(void *opaque, hwaddr addr, uint64_t val64,
+        size_t len = strlen(s->seed_xstrs[ix]);
+        size_t seed_len_bytes = KEYMGR_SEED_BYTES;
+        if (ix == KEYMGR_SEED_LFSR) {
+            seed_len_bytes = 8u;
```
Why 8? Please could you comment and/or constant-ify this?

AlexJones0 replied:
Thanks, that's a good spot. See https://github.com/lowRISC/opentitan/blob/75e1ba9a980f07243bc0d20a79e6bdabc91c3772/hw/ip/keymgr/data/keymgr.hjson#L153 compared to e.g. https://github.com/lowRISC/opentitan/blob/75e1ba9a980f07243bc0d20a79e6bdabc91c3772/hw/ip/keymgr/data/keymgr.hjson#L171 or any of the other `RndCnst.*Seed` keymgr parameters.

I've made this a constant `KEYMGR_LFSR_SEED_BYTES` as you suggest and added a small comment also to this effect.

jwnrt reviewed (approved):
Looks okay, thanks

luismarques commented:
Progresses https://github.com/lowRISC/opentitan/issues/27894

AlexJones0 committed:
[ot] scripts/opentitan: tests: Mark rsa_4096_signature_functest flaky

This test occasionally fails in CI due to crypto test timing checks
being quite tight. This is difficult to reproduce locally, so it may
even be due to differences in the active loads of the CI runners etc.

For now, mark this test as flaky to prevent issues in QEMU CI.

Signed-off-by: Alex Jones <alex.jones@lowrisc.org>

loiclefort commented on hw/opentitan/ot_keymgr.c:
```diff
@@ -83,6 +83,8 @@ static_assert(KEYMGR_GEN_DATA_BYTES <= KEYMGR_KDF_BUFFER_BYTES,
 #define KEYMGR_ENTROPY_WIDTH  (KEYMGR_LFSR_WIDTH / 2u)
 #define KEYMGR_ENTROPY_ROUNDS (KEYMGR_KEY_WIDTH / KEYMGR_ENTROPY_WIDTH)
 
+#define KEYMGR_LFSR_SEED_BYTES ((KEYMGR_LFSR_WIDTH) / 8u)
```
maybe add a static assert to make sure that `KEYMGR_LFSR_SEED_BYTES <= KEYMGR_SEED_BYTES`? (because `seeds` array is allocated with string of length `KEYMGR_SEED_BYTES`)

AlexJones0 replied:
Good idea, thanks - I've added that now.

loiclefort reviewed (approved):
Just a small suggestion to add a static_assert, otherwise LGTM
